client/my-sites/stats/features/modules/stats-utm/stats-module-utm.jsx

```diff
@@ -46,34 +46,68 @@ const StatsModuleUTM = ( {
 	query,
 	postId,
 	summaryUrl,
+	context,
 } ) => {
 	const siteId = useSelector( getSelectedSiteId );
 	const siteSlug = useSelector( ( state ) => getSiteSlug( state, siteId ) );
 	const translate = useTranslate();
 	const isOdysseyStats = config.isEnabled( 'is_running_in_jetpack_site' );
 	const [ selectedOption, setSelectedOption ] = useState( OPTION_KEYS.SOURCE_MEDIUM );
-	const [ url ] = useState( () => {
-		let initialUrl = summaryUrl;
+	const queryParams = useMemo( () => {
+		let urlParams;
 
-		if ( ! initialUrl ) {
-			initialUrl = isOdysseyStats ? window.location.hash.split( '#!' )[ 1 ] : window.location.href;
+		if ( summaryUrl ) {
+			urlParams = new URLSearchParams( summaryUrl?.split( '?' )[ 1 ] || '' );
+		} else {
+			urlParams = new URLSearchParams( context.query );
 		}
 
-		const queryParams = new URLSearchParams( initialUrl.split( '?' )[ 1 ] || '' );
+		return urlParams;
+	}, [ summary, summaryUrl, context.query ] );
+
+	const getUrlWithUpdatedParams = ( url ) => {
+		const currentUrl = new URL( url );
+		let updatedParams;
+
+		// Delete param in hash URL for Odyssey Stats if any.
+		if ( isOdysseyStats && currentUrl.hash.startsWith( '#!' ) ) {
+			const hashUrl = new URL( currentUrl.hash.substring( 2 ), currentUrl.origin );
+			hashUrl.searchParams.set( UTM_QUERY_PARAM, selectedOption );
+			currentUrl.hash = `#!${ hashUrl.pathname }${ hashUrl.search }`;
+			updatedParams = Object.fromEntries( hashUrl.searchParams.entries() );
+		} else {
+			currentUrl.searchParams.set( UTM_QUERY_PARAM, selectedOption );
+			updatedParams = Object.fromEntries( currentUrl.searchParams.entries() );
+		}
+		return { url: currentUrl, params: updatedParams };
+	};
+
+	useEffect( () => {
 		const utmParam = queryParams.get( UTM_QUERY_PARAM );
 
 		if ( utmParam && Object.values( OPTION_KEYS ).includes( utmParam ) ) {
 			setSelectedOption( utmParam );
 		}
-
-		return { url: initialUrl, params: queryParams };
-	} );
+	}, [ queryParams ] );
 
 	useEffect( () => {
-		if ( ! isOdysseyStats && summary ) {
-			const newUrl = new URL( window.location );
-			newUrl.searchParams.set( UTM_QUERY_PARAM, selectedOption );
-			window.history.replaceState( {}, '', newUrl );
+		if ( summary ) {
+			const newUrlObj = getUrlWithUpdatedParams( window.location.href );
```

Comment:
`page` should be able to deal with the occasion here if I'm not mistaken, like the following,

`page('/path/with/?param1=1&para2=2')`

Reply:
This does. Thank you for pointing me to this. This makes the code much simpler.